server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamRouteProvider.java

```diff
@@ -98,16 +98,34 @@ private String getKey(Deployment deployment) {
     private void evictExpiredBalancers() {
         long currentTime = System.currentTimeMillis();
         for (String key : balancers.keySet()) {
-            balancers.compute(key, (k, balancer) -> {
-                if (balancer != null && currentTime - balancer.getLastAccessTime() > IDLE_PERIOD_IN_MS) {
+            balancers.compute(key, (k, wrapper) -> {
+                if (wrapper != null && currentTime - wrapper.lastAccessTime > IDLE_PERIOD_IN_MS) {
                     return null;
                 }
-                return balancer;
+                return wrapper;
             });
         }
     }
 
     private static boolean isUpstreamsTheSame(List<Upstream> a, List<Upstream> b) {
         return new HashSet<>(a).equals(new HashSet<>(b));
     }
+
+    private static class BalancerWrapper {
+        final TieredBalancer balancer;
+        long lastAccessTime;
+
+        /**
+         * Note. The value is taken from {@link Deployment#getMaxRetryAttempts()} or {@link Route#getMaxRetryAttempts()}
+         */
+        final int maxRetryAttempts;
+
+        final List<Upstream> upstreams;
+
+        public BalancerWrapper(TieredBalancer balancer, int maxRetryAttempts, List<Upstream> upstreams) {
```

Comment:
You can simplify by taking just upstreams and maxRetryAttemps and create a balancer here.

Reply:
fixed